Native onchain development - Updated program security lesson (#452)
solana-foundation/developer-content: content/courses/native-onchain-development/program-security.md

---
title: Create a Basic Program, Part 3 - Basic Security and Validation
objectives:
  - Understand why "thinking like an attacker" is essential in securing Solana
    programs.
  - Learn and implement core security practices to protect your program.
  - Perform owner and signer checks to verify account ownership and transaction
    authenticity.
  - Validate the accounts passed into your program to ensure they are what you
    expect.
  - Conduct basic data validation to prevent invalid or malicious input from
    compromising your program.
description:
  "This lesson emphasizes securing your Solana program by implementing key
  validation checks. You'll learn how to perform ownership and signer checks,
  validate accounts to ensure they match your program's expectations, and

>   confirm that instruction data follows the required rules."

Reviewer: Sorry @0xCipherCoder this isn't your fault but this description is limited (just for courses) - see https://solana.com/developers/courses

> Learn the basics of how to interact with the Solana blockchain.

is a description. Make this about the size of those please!

Author: Updated description as per suggestion.

- **Data validation** verifies that the instruction data provided to your
  program adheres to specific rules or constraints, ensuring it doesn't lead to
  unintended behavior.

## Lesson

In the previous lessons
[deserialize instruction data](/content/courses/native-onchain-development/deserialize-instruction-data.md)
and
[program state management](/content/courses/native-onchain-development/program-state-management.md),
we built a Movie Review program, and while getting it to function was exciting,
secure development doesn't stop at "just working." It's critical to understand
potential failure points and take proactive steps to secure your program against
both accidental misuse and intentional exploitation.

Remember, **you have no control over the transactions that will be sent to your
program once it's deployed**. You can only control how your program handles
them. While this lesson is far from a comprehensive overview of program
security, we'll cover some of the basic pitfalls to look out for.

### Think Like an Attacker

A fundamental principle in secure programming is adopting an "attacker's
mindset." This means considering every possible angle someone might use to break

> or exploit your program, especially in scenarios where you may not have direct
> control over how users or third parties interact with your deployed code.

Reviewer: ```suggestion
or exploit your program.
```

I like your other edits but this might be unintentionally misleading:

>  scenarios where you may not have direct control over how users or third parties interact with your deployed code.

I don't think this is what you meant, but the way this is written is that we sometimes control other people - we do not control how users or third parties interact with our code ever.

Author: Removed this to avoid confusion.

Every program has vulnerabilities. The question isn't whether it can be broken,
but how much effort it takes. As developers, our goal is to close as many
security gaps as possible and increase the effort required to break our code.
For example, while our Movie Review program creates accounts to store reviews,
there may be unintentional behaviors that could be caught by thinking like an
attacker. In this lesson, we'll explore these issues and how to address them.

### Error handling

Before we dive into some of the common security pitfalls and how to avoid them,
it's important to know how to use errors in your program. Security issues in a
Solana program often requires terminating the execution with a meaningful error.
Not all errors are catastrophic, but some should result in stopping the program
and returning an appropriate error code to prevent further processing.

#### Creating Custom Errors

Solana's
[`solana_program`](https://docs.rs/solana-program/latest/solana_program/) crate
provides a generic
[`ProgramError`](https://docs.rs/solana-program/latest/solana_program/program_error/enum.ProgramError.html)
enum for error handling. However, custom errors allow you to provide more

> detailed, context-specific information that helps during debugging and testing.

Reviewer: Love the links. 👍

We can define our own errors by creating an enum type listing the errors we want
to use. For example, the `NoteError` contains variants `Forbidden` and
`InvalidLength`. The enum is made into a Rust `Error` type by using the `derive`
attribute macro to implement the `Error` trait from the
[`thiserror`](https://docs.rs/thiserror/latest/thiserror/) library. Each error
type also has its own `#[error("...")]` notation. This lets you provide an error
message for each particular error type.

Here's an example of how you can define custom errors in your program:

```rust
use solana_program::program_error::ProgramError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum NoteError {
    #[error("Unauthorized access - You don't own this note.")]
    Forbidden,

    #[error("Invalid note length - The text exceeds the allowed limit.")]
    InvalidLength,
}
```

> <Callout>
> 
> In this example, we create custom errors for unauthorized access and invalid
> data input (such as note length). Defining custom errors gives us greater
> flexibility when debugging or explaining what went wrong during execution.
> </Callout>

Reviewer: ```suggestion
```

Yes good errors help us debug, that should be obvious. I'm not sure this warrants a callout.

Author: Removed callout as it's not that important.

}
```

This ensures the program gracefully handles errors and provides meaningful
feedback when things go wrong.

### Basic Security Checks

To ensure your Solana program is resilient against common vulnerabilities, you
should incorporate key security checks. These are critical for detecting invalid
accounts or unauthorized transactions and preventing undesired behavior.

#### Ownership checks

An ownership check verifies that an account is owned by the expected program.
For instance, if your program relies on PDAs (Program Derived Addresses), you
want to ensure that those PDAs are controlled by your program and not by an
external party.

> Let's use the note-taking app example that we've referenced in previous lessons
> [deserialize instruction data](/content/courses/native-onchain-development/deserialize-instruction-data.md)
> and
> [program state management](/content/courses/native-onchain-development/program-state-management.md).

Reviewer: ```suggestion
Let's use the note-taking app example that we've referenced in the 
[deserialize instruction data](/content/courses/native-onchain-development/deserialize-instruction-data.md)
and
[program state management](/content/courses/native-onchain-development/program-state-management.md) lessons.
```

Author: Updated description as per suggestion.

program in PDA accounts.

When a user invokes the `update` instruction handler, they also provide a
`pda_account`. We presume the provided `pda_account` is for the particular note
they want to update, but the user can input any instruction data they want. They
could even potentially send data that matches the data format of a note account
but was not also created by the note-taking program. This security vulnerability
is one potential way to introduce malicious code.

The simplest way to avoid this problem is to always check that the owner of an
account is the public key you expect it to be. In this case, we expect the note
account to be a PDA account owned by the program itself. When this is not the
case, we can report it as an error accordingly.

Here's how you can perform an ownership check to verify that an account is owned
by the program:

```rust
if note_pda.owner != program_id {
    return Err(ProgramError::InvalidNoteAccount);
}
```

In this example, we check if the `note_pda` is owned by the program itself
(denoted by `program_id`). Ownership checks like these prevent unauthorized
entities from tampering with critical accounts.

<Callout>

> PDAs, in particular, are more secure than externally-owned accounts because the
> program has full control over them. Ensuring ownership is a fundamental way to
> prevent malicious behavior. </Callout>

Reviewer: ```suggestion
PDAs are often considered to be trusted stores of a program's state. Ensuring the correct program owns the PDAs is a fundamental way to prevent malicious behavior. </Callout>
```

Reviewer: Making it simpler...

Author: Updated as per suggestion.

}
```

By verifying that the initializer has signed the transaction, we ensure that
only the legitimate owner of the account can perform actions on it.

#### Account Validation

Account validation checks that the accounts passed into the program are correct
and valid. This is often done by deriving the expected account using known seeds
(for PDAs) and comparing it to the passed account.

For instance, in the note-taking app, you can derive the expected PDA using the
creator's public key and note ID, and then validate that it matches the provided
account:

```rust
let (expected_pda, bump_seed) = Pubkey::find_program_address(
    &[
        note_creator.key.as_ref(),
        id.as_bytes().as_ref(),
    ],
    program_id
);

> if expected_pda != *note_pda.key {

Reviewer: Your expected/actual style is much better 👍

#### Integer overflow and underflow

One of the common pitfalls when working with integers in Rust (and in Solana
programs) is handling integer overflow and underflow. Rust integers have fixed
sizes and can only hold values within a certain range. When a value exceeds that
range, it wraps around, leading to unexpected results.

For example, with a `u8` (which holds values between 0 and 255), adding 1 to 255
results in a value of 0 (overflow). To avoid this, you should use checked math
functions like
[`checked_add()`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_add)
and

> [`checked_sub()`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_sub):

Reviewer: links 👍

To avoid integer overflow and underflow, either:

1. Have logic in place that ensures overflow or underflow _cannot_ happen or
2. Use checked math like `checked_add()` instead of `+`

   ```rust
   let first_int: u8 = 5;
   let second_int: u8 = 255;
   let sum = first_int.checked_add(second_int)
    .ok_or(ProgramError::ArithmeticOverflow)?;
   ```

## Lab

> In this lab, we will build upon the Movie Review program that allows users to

Reviewer: This was in the old lesson too, but Github markdown seems to break here - syntax highlighting starts including 'in' and 'self' etc as keywords in regular text. Maybe check the previous code block is terminated properly.

Author: I tried to find if there is anything missing or breaking in markdown but I am not able to find any issue and even quotes are fine. Let me know what can be done and this is not getting reflected in the rich diff view.
